javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -0,0 +1,205 @@
+package java.util.concurrent
+
+import scala.collection.JavaConversions._
+
+trait ConcurrentMap[K, V] extends java.util.Map[K, V] {
+
+  def remove(key: Any, value: Any): Boolean = {
+    if (containsKey(key) && get(key).equals(value)) {
+      remove(key)
+      true
+    } else false
+  }
+
+  def replace(key: K, value: V): V = {
+    if (containsKey(key)) {
+      put(key, value)
+    } else get(key)
+  }
+
+  def replace(key: K, oldValue: V, newValue: V): Boolean = {
+    if (containsKey(key) && get(key).equals(oldValue)) {
+      put(key, newValue)
+      true
+    } else false
+  }
+}
+
+class ConcurrentHashMap[K >: Null, V >: Null](initialCapacity: Int,
+    loadFactor: Float,
+    concurrencyLevel: Int,
+    initialMap: java.util.Map[_ <: K, _ <: V])
+    extends java.util.AbstractMap[K, V]
+    with java.util.concurrent.ConcurrentMap[K, V]
+    with java.io.Serializable {
+
+  def this() = 
+    this(16, 0.75F, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int) = 
+    this(initialCapacity, 0.75F, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int, loadFactor: Float) = 
+    this(initialCapacity, loadFactor, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int, loadFactor: Float, concurrencyLevel: Int) = 
+    this(initialCapacity, loadFactor, concurrencyLevel, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = 
+    this(16, 0.75F, 16, initialMap)
+
+  putAll(initialMap)
+
+  private lazy val inner = new scala.collection.mutable.HashMap[K, V]()
+
+  override def clear(): Unit =
+    inner.clear
+
+  override def containsKey(key: Any): Boolean =
+    inner.contains(key.asInstanceOf[K])
```

Comment:
Ah ... This is annoying. I don't think it's possible to guard against `key` not being a `K`, because `K` is erased at this point.

I have thought some more about this, and I don't see how casting to `K` right here could be a problem in practice, given that we know that we're working with a `HashMap`, which is entirely generic. So just keep what you wrote (`key.asInstanceOf[K]`). I don't think it's possible to write a test that would break it.